include/proxy-wasm/context_interface.h

```diff
@@ -0,0 +1,628 @@
+   * WasmHeaderMapType::GrpcReceiveTrailingMetadata as the type.
+   */
+  virtual void onGrpcReceiveTrailingMetadata(GrpcToken token) = 0;
+
+  /**
+   * Called on Root Context when a gRPC call has closed prematurely or a gRPC stream has closed.
+   * @token is the token returned by grpcStream.
+   * @status is status of the stream.
+   * @message is any status message.
+   * For Call only used when not Ok. For stream indicates the peer has closed.
+   */
+  virtual void onGrpcClose(GrpcToken token, const GrpcStatus &status,
+                           const string_view message) = 0;
+};
+
+/**
+ * RootInterface is the interface specific to RootContexts.
+ * A RootContext is associated with one more more plugins and is the parent of all stream Context(s)
+ * created for that plugin. It can be used to store data shared between stream Context(s) in the
+ * same VM.
+ */
+struct RootInterface : public RootGrpcInterface {
+  /**
+   * Call on a host Context to create a corresponding Context in the VM.  Note:
+   * onNetworkNewConnection and onRequestHeaders() call onCreate().
+   * @param parent_context_id is the parent Context id for the context being created.  For a
+   * stream Context this will be a Root Context id (or sub-Context thereof).
+   */
+  virtual void onCreate(uint32_t parent_context_id);
+
+  /**
+   * Call on a Root Context when a VM first starts up.
+   * @param plugin is the plugin which caused the VM to be started.
+   * Called by the host code.
+   */
+  virtual bool onStart(std::shared_ptr<PluginBase> plugin) = 0;
+
+  /**
+   * Call on a Root Context when a plugin is configured or reconfigured dynamically.
+   * @param plugin is the plugin which is being configured or reconfigured.
+   */
+  virtual bool onConfigure(std::shared_ptr<PluginBase> plugin) = 0;
+
+  /**
+   * Called on a Root Context when a response arrives for an outstanding httpCall().
+   * @param token is the token returned by the corresponding httpCall().
+   */
+  virtual void onHttpCallResponse(HttpCallToken token) = 0;
+
+  /**
+   * Called on a Root Context when an Inter-VM shared queue message has arrived.
+   * @token is the token returned by registerSharedQueue().
+   */
+  virtual void onQueueReady(uint32_t SharedQueueDequeueToken) = 0;
+
+  /**
+   * Call when a stream has completed (both sides have closed) or on a Root Context when the VM is
+   * shutting down.
+   * @return true for stream contexts or for Root Context(s) if the VM can shutdown, false for Root
+   * Context(s) if the VM should wait until the Root Context calls the proxy_done() ABI call.  Note:
+   * the VM may (optionally) shutdown after some configured timeout even if the Root Context does
+   * not call proxy_done().
+   */
+  virtual bool onDone() = 0;
+
+  /**
+   * Call when no further stream calls will occur.  This will cause the corresponding Context in the
+   * VM to be deleted.
+   * Called by the host code.
+   */
+  virtual void onDelete();
```

Comment:
Why this method is not pure virtual like the rest ?

Reply:
Fixed